### ISSUE TITLE
/dor mount not reliably cleared out by accessionWF reset-workspace

### ISSUE THREAD

**Issue by tallenaz** [sul-dlss/common-accessioning#321]
Is this a dup of #342 ?  Is the fix similar?  Can we fix #342 at the same time, please?

<hr />

While debugging https://jirasul.stanford.edu/jira/browse/SDRO-266 @blalbrit and I discovered that the dor assembly directory is not cleaned out by reset-workspace as reliably as we expect.

**Comment by blalbrit** [comment]
The directories that should be cleaned out will be:
* /dor/export
* /dor/workspace
* /dor/assembly

**Comment by jmartin-sul** [comment]
might errors like this be related?
https://app.honeybadger.io/projects/52894/faults/49769895

`Errno::ENOENT: No such file or directory @ rb_file_s_rename - (/dor/workspace/zy/501/np/2944/zy501np2944, /dor/workspace/zy/501/np/2944/zy501np2944_v1)`
```
lib/reset_workspace_service.rb:11:in `reset_workspace_druid_tree`
lib/robots/dor_repo/accession/reset_workspace.rb:18:in `perform`
```

or is that a different problem?

if the latter, i can file a separate issue for figuring out whether there's a bug, or just a noisy alert.

**Comment by jmartin-sul** [comment]
same question about e.g.:
https://app.honeybadger.io/projects/52894/faults/50132373

`Errno::ENOENT: No such file or directory @ dir_initialize - /dor/workspace/zy/501/np/2944`

```
lib/cleanup_reset_service.rb:39:in `cleanup_reset_workspace_content`
lib/cleanup_reset_service.rb:11:in `cleanup_by_reset_druid`
lib/robots/dor_repo/dissemination/cleanup.rb:14:in `perform`
```

**Comment by peetucket** [comment]
Seems to be ongoing (new files with datestamps of May - July 2019)

**Comment by peetucket** [comment]
Impact: If the /dor mount fills up over time due to the /dor/export and /dor/workspace not being cleared after accessioning/ingest is complete, a nagios alert goes out.  This requires some manual intervention to investigate and remove content and possibly some scripting to be sure we don't inadvertently remove content still in accessioning or versioning.  So the impact is mostly operations and developer time spent periodically tending to the drive filling up when it alerts.

There is no user facing impact, other than the possible pausing of accessioning if the /dor mount were to be filled up and not cleared manually.

**Comment by ndushay** [comment]
I'm treating this like a duplicate of #342.  Will add note to #342 to address this ticket, too.

**Comment by ndushay** [comment]
from #342:

- what does `accessionWF:reset-workspace` does as opposed to `disseminationWF:cleanup`?
  - `accessionWF:reset-workspace` moves the `/dor/workspace` druid tree folder to an archived folder in the base of the druid tree in `/dor/workspace`
- Still not sure what clears out the archive folder though (e.g. /dor/workspace/zz/998/dj/6963/zz998dj6963_v2)
- what clears out `/dor/assembly` ?

**Comment by ndushay** [comment]
See also argo errors and mebbe honeybadger too???

**Comment by ndushay** [comment]
See #333 also - we should use the new way to talk to dor-services.

**Comment by mjgiarlo** [comment]
Possibly though maybe not related is this dor-services-app alert: https://app.honeybadger.io/projects/50568/faults/53522098

The problem in this class is when this method raises:

https://github.com/sul-dlss/dor-services-app/blob/master/app/services/cleanup_reset_service.rb#L11

In which case, these two methods are not called: https://github.com/sul-dlss/dor-services-app/blob/master/app/services/cleanup_reset_service.rb#L12-L13

**Comment by peetucket** [comment]
Ok, that was interesting lead Mike.  But based on a recent factor that moved the pruning from druid-tools to a new `PruneService` class in dor-services-app, I don't think we will see the https://app.honeybadger.io/projects/50568/faults/53522098 exception anymore.  This is because `druid-tools` used to use the File-Utils method `#rmdir` while the PruneService now uses the File-Utils method `#rmtree`.  `#rmdir` throws an exception if the folder you are trying to delete is not there (as shown in the HB exception above).  On the other hanb, `rmtree` is aliased to `rm_rf` which itself is equivalent to `rm_r` with the :force option set to true.  When :force is set to true, missing files or folders do not cause an exception.

I think is ok -- the fact that a folder in the druid tree hierarchy is gone is ok, we really just want it to be deleted, and if a race condition has caused another druid with similar druid tree hiearchy to clean it up ahead of time, that's fine.

**Comment by peetucket** [comment]
What we know so far:

The following robots do the following things, in this order:

1. `accessionWF:reset-workspace` : This robot uses the `ResetWorkspaceService` defined in common-accessioning to:
* reset_workspace_druid_tree : moves the latest workspace druid tree to a folder based on the version number, e.g. for `druid:oo000oo0001` that is currently version 2, it would move `/dor/workspace/oo/000/oo/0001/oo000oo0001` to `/dor/workspace/oo/000/oo/0001/oo000oo0001_v2`.  If the version folder already exists (e.g. `/dor/workspace/oo/000/oo/0001/oo000oo0001_v2`) the robot will go into an error state.  If the druid tree folder (e.g. `/dor/workspace/oo/000/oo/0001/oo000oo0001`) doesn't exist for some reason, nothing happens (but the step will complete ok).
* reset_export_bag : moves the current druid export directory to a folder based on version number, e.g. for `druid:oo000oo0001` that is currently version 2, it would move `/dor/export/oo000oo0001` to `/dor/export/oo000oo0001_v2` and it would also move a similarly named .tar file: `/dor/export/oo000oo0001.tar` to `/dor/export/oo000oo0001_v2.tar`.  If the version folder already exists (.eg. `/dor/export/oo000oo0001_v2`) the robot will go into an error state.  If the export folders or tars do not exist, nothing happens (but the step will complete ok).

So in summary, nothing is deleted in `accessionWF:reset-workspace` ... just moving/renaming folders.

**Comment by justinlittman** [comment]
For what it is worth, if this is on an NFS mount, it isn't surprising that these sorts of file operations (mv, rm) are hinky.

**Comment by peetucket** [comment]
I have done some analysis, and I think I now have a better idea of where stuff should be deleted. I don't many theories about why it sometimes doesn't seem to work, but some ideas of what we can try next.

**Comment by andrewjbtw** [comment]
I think the only actionable work may be connected to: https://github.com/sul-dlss/argo/issues/1878

We're definitely cleaning up /dor/assembly and I think /dor/workspace is also being cleaned up.

**Comment by andrewjbtw** [comment]
I think where we stand on this in 2022 is:

- reset-workspace reliably clears out the appropriate /dor/assembly and /dor/workspace folders when the step runs without an error
- reset-workspace does not clear out the appropriate /dor/export folder (see https://github.com/sul-dlss/argo/issues/1878 also linked above)
- it's not clear if files are cleaned up anywhere if reset-workspace runs into an error and is then retried

**Comment by andrewjbtw** [comment]
I'm also happy to create a new ticket because the workflows are a bit different now. There's no disseminationWF and I think the deletion was moved up to end-accession but am not 100% clear.

**Comment by ndushay** [comment]
Andrew thinks this is done, but it needs to be verified.

**Comment by ndushay** [comment]
@jcoyne assigning this to you because your PR https://github.com/sul-dlss/dor-services-app/pull/770 addresses this ticket as well.
